src/rules/no-duplicated-branches.ts

```diff
@@ -0,0 +1,119 @@
+import { Rule } from "eslint";
+import * as estree from "estree";
+import { getParent, isIfStatement, isBlockStatement } from "../utils/nodes";
+import { areEquivalent } from "../utils/equivalence";
+
+const MESSAGE = "This {{type}}'s code block is the same as the block for the {{type}} on line {{line}}.";
+
+const rule: Rule.RuleModule = {
+  create(context: Rule.RuleContext) {
+    return {
+      IfStatement(node: estree.Node) {
+        visitIfStatement(node as estree.IfStatement);
+      },
+      SwitchStatement(node: estree.Node) {
+        visitSwitchStatement(node as estree.SwitchStatement);
+      },
+    };
+
+    function visitIfStatement(ifStmt: estree.IfStatement) {
+      const parent = getParent(context);
+      if (!isIfStatement(parent)) {
+        const branches = collectIfBranches(ifStmt);
+
+        for (let i = 1; i < branches.length; i++) {
+          if (hasRequiredSize([branches[i]])) {
+            for (let j = 0; j < i; j++) {
+              if (compareIfBranches(branches[i], branches[j])) {
+                break;
+              }
+            }
+          }
+        }
+      }
+    }
+
+    function visitSwitchStatement({ cases }: estree.SwitchStatement) {
+      for (let i = 1; i < cases.length; i++) {
+        const firstClauseWithoutBreak = takeWithoutBreak(expandSingleBlockStatement(cases[i].consequent));
+
+        if (hasRequiredSize(firstClauseWithoutBreak)) {
+          for (let j = 0; j < i; j++) {
+            const secondClauseWithoutBreak = takeWithoutBreak(expandSingleBlockStatement(cases[j].consequent));
+
+            if (areEquivalent(firstClauseWithoutBreak, secondClauseWithoutBreak, context.getSourceCode())) {
```

Comment:
@stas-vilchik I like the way you did with `compareIfBranches`, much more readable. Can we refactor the same way here?